beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	numOfVals := state.NumValidators()
+	// Guard against an out-of-bounds using validator balance precompute.
+	if len(vals) != numOfVals || len(vals) != state.BalancesLength() {
+		return state, errors.New("registries not the same length as state registries")
+	}
+
+	attsRewards, attsPenalties, err := AttestationsDelta(state, bal, vals)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get attestation delta")
+	}
+
+	balances := state.Balances()
+	for i := 0; i < numOfVals; i++ {
+		vals[i].BeforeEpochTransitionBalance = balances[i]
+
+		// Compute the post balance of the validator after accounting for the
+		// attester and proposer rewards and penalties.
+		balances[i] = helpers.IncreaseBalanceWithVal(balances[i], attsRewards[i])
+		balances[i] = helpers.DecreaseBalanceWithVal(balances[i], attsPenalties[i])
+
+		vals[i].AfterEpochTransitionBalance = balances[i]
+	}
+
+	if err := state.SetBalances(balances); err != nil {
+		return nil, errors.Wrap(err, "could not set validator balances")
+	}
+
+	return state, nil
+}
+
+// AttestationsDelta computes and returns the rewards and penalties differences for individual validators based on the
+// voting records.
+func AttestationsDelta(state iface.BeaconStateAltair, bal *precompute.Balance, vals []*precompute.Validator) ([]uint64, []uint64, error) {
+	numOfVals := state.NumValidators()
+	rewards := make([]uint64, numOfVals)
+	penalties := make([]uint64, numOfVals)
+	prevEpoch := helpers.PrevEpoch(state)
+	finalizedEpoch := state.FinalizedCheckpointEpoch()
+
+	for i, v := range vals {
+		rewards[i], penalties[i] = attestationDelta(bal, v, prevEpoch, finalizedEpoch)
+	}
+	return rewards, penalties, nil
+}
+
+func attestationDelta(bal *precompute.Balance, v *precompute.Validator, prevEpoch, finalizedEpoch types.Epoch) (uint64, uint64) {
+	eligible := v.IsActivePrevEpoch || (v.IsSlashed && !v.IsWithdrawableCurrentEpoch)
+	if !eligible || bal.ActiveCurrentEpoch == 0 {
+		return 0, 0
+	}
+
+	ebi := params.BeaconConfig().EffectiveBalanceIncrement
+	eb := v.CurrentEpochEffectiveBalance
+	br := eb * params.BeaconConfig().BaseRewardFactor / mathutil.IntegerSquareRoot(bal.ActiveCurrentEpoch)
+	activeCurrentEpochIncrements := bal.ActiveCurrentEpoch / ebi
+
+	r, p := uint64(0), uint64(0)
+	// Process source reward / penalty
+	if v.IsPrevEpochAttester && !v.IsSlashed {
+		if helpers.IsInInactivityLeak(prevEpoch, finalizedEpoch) {
+			// Since full base reward will be canceled out by inactivity penalty deltas,
+			// optimal participation receives full base reward compensation here.
+			r += br * params.BeaconConfig().TimelySourceWeight / params.BeaconConfig().WeightDenominator
+		} else {
+			rewardNumerator := br * params.BeaconConfig().TimelySourceWeight * (bal.PrevEpochAttested / ebi)
+			r += rewardNumerator / (activeCurrentEpochIncrements * params.BeaconConfig().WeightDenominator)
+		}
+	} else {
+		p += br * params.BeaconConfig().TimelySourceWeight / params.BeaconConfig().WeightDenominator
+	}
```

Comment:
I think three blocks (process source/target/head rewards/penalties) are doing almost identical thing, for different flags -- thus can be extracted into a single method. This will make the method less complex, and review easier.

It is probably a good idea to copy into comments for that extracted method, the [get_flag_index_deltas()](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/altair/beacon-chain.md#get_flag_index_deltas) specs chunk for reference (I suppose the computation implements that specs part).

Reply:
There's a subtle difference on why we want to do it this way. We want different `unslashed_participating_increments` for every flag index. The spec chooses readability but sacrificed it with O(n^2) run time where this approach is O(n). We did this for Phase 0 too

Reply:
Understood, thank you for elaboration.